### PR TITLE
Add support for NDK r19 prebuilt toolchains

### PR DESCRIPTION
**Pull request by str4d**
This skips the `generateToolchains` task if it is unnecessary.

### CODE REVIEW

**Comment by str4d** [comment]
> why is the generateToolchains task conditional on the NDK revision?

Apologies, I should have been more verbose in the initial PR. Since r19, the Android NDK provides prebuilt toolchains for each of the four supported targets, and `make_standalone_toolchain.py` is no longer necessary. See the warning at the top of the [NDK Standalone Toolchains guide](https://developer.android.com/ndk/guides/standalone_toolchain.html), and the new ["other build systems" guide](https://developer.android.com/ndk/guides/other_build_systems). I originally just removed `generateToolchains` entirely, but then decided to instead have the plugin support both in case you wanted to maintain backwards-compatibility with pre-r19 NDKs. If that isn't desired, this PR becomes significantly simpler (and the global flag goes away).

This change reduces disk usage during builds (as the toolchains aren't being duplicated), and removes one of the two dependencies on Python (the other being the Cargo linker wrapping introduced in #21).

**Comment by ncalexan** [comment]
> > why is the generateToolchains task conditional on the NDK revision?
> 
> Apologies, I should have been more verbose in the initial PR. Since r19, the Android NDK provides prebuilt toolchains for each of the four supported targets, and `make_standalone_toolchain.py` is no longer necessary. See the warning at the top of the [NDK Standalone Toolchains guide](https://developer.android.com/ndk/guides/standalone_toolchain.html), and the new ["other build systems" guide](https://developer.android.com/ndk/guides/other_build_systems). I originally just removed `generateToolchains` entirely, but then decided to instead have the plugin support both in case you wanted to maintain backwards-compatibility with pre-r19 NDKs. If that isn't desired, this PR becomes significantly simpler (and the global flag goes away).

Mmm, right.  In the Firefox tree we've had endless problems with the pre-built NDKs, so much so that we build our own clang toolchains now, but that's very unlikely to be relevant to folks using this plugin.

> This change reduces disk usage during builds for NDK r19+ (as the toolchains aren't being duplicated), and removes one of the two dependencies on Python (the other being the Cargo linker wrapping introduced in #21).

Roger that.  So, rather than making the flag be "r19", let's let the user choose whether to use standalone toolchains.  Expose that flag in the extension DSL but default it to true if we detect r19+.  I still think the cleanest way to do this is to have the `Toolchain` definitions know if they're using a standalone or if they're using a prebuilt, so can you have a cut at that expression?

Thanks, @str4d!

**Comment by str4d** [comment]
I've renamed the `r19` boolean to `prebuilt` in the original commit, and exposed it in the DSL. I'll next try experimenting with changes to the `Toolchain` definitions.

**Comment by str4d** [comment]
@ncalexan I've pushed commits to switch to using separate `ToolchainType`s instead of a `prebuilt` flag being passed through to `CargoBuildTask` and `Toolchain`. I'm not convinced it's a better design, but it works (at least with me testing `prebuiltToolchains = true|false` in the sample library). Let me know what you think.

File: plugin/src/main/kotlin/com/nishtahir/RustAndroidPlugin.kt [hunk 1]
```diff
@@ -20,67 +21,104 @@ val toolchains = listOf(
...
         Toolchain("x86_64",
-                ToolchainType.ANDROID,
+                ToolchainType.ANDROID_GENERATED,
+                "x86_64-linux-android",
+                "x86_64-linux-android",
+                "x86_64-linux-android",
+                "android/x86_64"),
+        Toolchain("arm",
+                ToolchainType.ANDROID_PREBUILT,
+                "armv7-linux-androideabi",
```
Comment (ncalexan) [inline plugin/src/main/kotlin/com/nishtahir/RustAndroidPlugin.kt]: Are you confident this is correct?  We have `armv7-linux-androideabi` for the Rust target -- that seems right.  But then the compiler and binutils triples differ?  That seems odd to to me, but I don't have an r19+ NDK to crack open right this moment.  Definitely r17b has `arm-linux-androideabi` for both, and you have that for the generated type above.

Comment (str4d) [reply plugin/src/main/kotlin/com/nishtahir/RustAndroidPlugin.kt]: See the note in [this section](https://developer.android.com/ndk/guides/other_build_systems#overview) about this. I agree it's weird, and it's annoying that it means we have three different target strings for prebuilt armv7, but that's the way it is.

**Comment by str4d** [comment]
Also, FYI I have several subsequent changes I'm considering making PRs here for:

- A `forceTargets` parameter that enables using 64-bit targets with pre-21 `minApi` ([commit](https://github.com/str4d/rust-android-gradle/commit/3ac008592a45d147bfc0302d72983c6c51640648))
  - According to our Android developer, it is common in recent large enterprise applications to target NDK API 16 for 32-bit targets, and API 21 for 64-bit targets. Thus it is currently necessary for us to support this configuration.
- A fix for part of #29 ([commit](https://github.com/str4d/rust-android-gradle/commit/a7847cd48d28ebdf1005370a663bccf91204d77a)).
- #23 Variant support (I haven't implemented this yet, but I require it for our library).

So you may not want to immediately cut a new release, or you may want to cut one anyway and leave these to subsequent releases.

**Pull request closed without merging**